src/audio/eq_iir.c

```diff
@@ -143,99 +130,128 @@ static void eq_iir_free_delaylines(struct iir_state_df2t *iir)
 	 * allocated buffer
 	 */
 	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++) {
-		if ((iir[i].delay != NULL) && (delay == NULL))
+		if (iir[i].delay && !delay)
 			delay = iir[i].delay;
 
 		/* Point all delays to NULL to avoid duplicated free later */
 		iir[i].delay = NULL;
 	}
 
-	if (delay != NULL)
+	if (delay)
 		rfree(delay);
-
 }
 
 static int eq_iir_setup(struct iir_state_df2t iir[],
-	struct sof_eq_iir_config *config, int nch)
+			struct sof_eq_iir_config *config, int nch)
 {
+	struct sof_eq_iir_header_df2t *lookup[SOF_EQ_IIR_MAX_RESPONSES];
+	struct sof_eq_iir_header_df2t *eq;
+	int64_t *iir_delay;
+	int32_t *coef_data, *assign_response;
+	size_t s;
+	size_t size_sum = 0;
 	int i;
 	int j;
-	int idx;
 	int resp;
-	size_t s;
-	size_t size_sum = 0;
-	int64_t *iir_delay; /* TODO should not need to know the type */
-	int32_t *coef_data, *assign_response;
-	int response_index[PLATFORM_MAX_CHANNELS];
 
 	/* Free existing IIR channels data if it was allocated */
 	eq_iir_free_delaylines(iir);
 
-	if ((nch > PLATFORM_MAX_CHANNELS)
-		|| (config->channels_in_config > PLATFORM_MAX_CHANNELS))
+	trace_eq("fse");
+	trace_value(config->channels_in_config);
+	trace_value(config->number_of_responses);
+
+	/* Sanity checks */
+	if (nch > PLATFORM_MAX_CHANNELS ||
+	    config->channels_in_config > PLATFORM_MAX_CHANNELS) {
+		trace_eq_error("ech");
 		return -EINVAL;
+	}
+	if (config->number_of_responses > SOF_EQ_IIR_MAX_RESPONSES) {
+		trace_eq_error("enr");
+		return -EINVAL;
+	}
 
 	/* Collect index of response start positions in all_coefficients[]  */
 	j = 0;
 	assign_response = &config->data[0];
 	coef_data = &config->data[config->channels_in_config];
-	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++) {
+	trace_eq("idx");
+	for (i = 0; i < SOF_EQ_IIR_MAX_RESPONSES; i++) {
 		if (i < config->number_of_responses) {
-			response_index[i] = j;
-			j += NHEADER_DF2T
-				+ NBIQUAD_DF2T * coef_data[j];
+			eq = (struct sof_eq_iir_header_df2t *)&coef_data[j];
+			lookup[i] = eq;
+			trace_value(j);
```

Comment:
might be more logical to trace before using the j index?

Reply:
yep